Feature/new executions

kuegi opened the pull request:
switched handling of executions (not only full order, but also partial fills)

also fixed backtestengine

kuegi committed:
[Engine, Bot] now directly reacting to executions

kuegi committed:
[Bot] fixes and format

kuegi committed:
improved position

added customData to store simple data to the position, also added last_entry field

kuegi committed:
[bybit-linear] fix reconnect on public stream

kuegi committed:
[MultiStrat] added typing

kuegi committed:
[trading_bot] when a pos goes to 0 size: mark as closed

even if it wasn't filled before.

kuegi committed:
fixed on_execution errros

kuegi committed:
bybit linear: fixed API type f* up

also improved order logging, and execution handling

kuegi committed:
[backtest] made executions more realistic in backtest

fixed new execution handling in live too

Krysh90 commented on kuegi_bot/backtest_engine.py:
```diff
@@ -175,13 +177,23 @@ def orderKeyForSort(self,order):
-        something_changed = False
+    def check_executions(self, intrabarToCheck: Bar, onlyOnClose):
         another_round= True
+        didSomething= False
```
rename to did_something

Krysh90 commented on kuegi_bot/backtest_engine.py:
```diff
@@ -175,13 +177,23 @@ def orderKeyForSort(self,order):
+    def check_executions(self, intrabarToCheck: Bar, onlyOnClose):
         another_round= True
+        didSomething= False
+        allowedOrderIds= None
```
allowed_order_ids

Krysh90 commented on kuegi_bot/backtest_engine.py:
```diff
@@ -198,17 +210,32 @@ def handle_open_orders(self, intrabarToCheck: Bar) -> bool:
                     break
+        return didSomething
+
+    def handle_subbar(self, intrabarToCheck: Bar):
```
intrabar_to_check

Krysh90 replied:
resolved but not changed? 🤔  i mean yeah it is only code-styling 🤷

Krysh90 commented on kuegi_bot/backtest_engine.py:
```diff
@@ -198,17 +210,32 @@ def handle_open_orders(self, intrabarToCheck: Bar) -> bool:
+        didSomething1= self.check_executions(intrabarToCheck,False)
+        self.current_bars[0].add_subbar(intrabarToCheck)  # so bot knows about the current intrabar
+
+        didSomething2= self.check_executions(intrabarToCheck,True)
```
didSomething1 and 2? is there no more meaningful meaning for those?

anyways would be did_something_1 and did_something_2

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -39,7 +39,7 @@ def __init__(self, logger, directionFilter: int = 0):
         self.time_of_max_equity = 0
         self.position_history: List[Position] = []
         self.openPositionRolling = 1
```
open_position_rolling

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -39,7 +39,7 @@ def __init__(self, logger, directionFilter: int = 0):
         self.position_history: List[Position] = []
         self.openPositionRolling = 1
-        self.unaccountedPositionCoolOff= 0
+        self.unaccountedPositionCoolOff = 0
```
unaccounted_position_cool_off

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -171,15 +171,61 @@ def position_got_opened(self, position: Position, bars: List[Bar], account: Acco
         self.position_got_opened(position, bars, account)
 
+    def on_execution(self, orderId, amount, executed_price, tstamp):
+        posId = self.position_id_from_order_id(orderId)
```
pos_id

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -171,15 +171,61 @@ def position_got_opened(self, position: Position, bars: List[Bar], account: Acco
-
         self.position_got_opened(position, bars, account)
 
+    def on_execution(self, orderId, amount, executed_price, tstamp):
```
order_id

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -171,15 +171,61 @@ def position_got_opened(self, position: Position, bars: List[Bar], account: Acco
+            return
+        position: Position = self.open_positions[posId]
+
+        orderType = self.order_type_from_order_id(orderId)
```
order_type

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -171,15 +171,61 @@ def position_got_opened(self, position: Position, bars: List[Bar], account: Acco
+        orderType = self.order_type_from_order_id(orderId)
+        position.changed = True
+        if orderType == OrderType.ENTRY:
+            position.status = PositionStatus.OPEN
```
missing logger.info for position got opened?

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -188,26 +234,22 @@ def sync_executions(self, bars: List[Bar], account: Account):
-                    position.exit_tstamp = order.execution_tstamp
-                    position.exit_equity = account.equity
-                    self.position_closed(position, account)
-                else:
```
this removed part has been moved to on_execution - right?

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -443,7 +487,7 @@ def save_open_positions(self, bars: List[Bar]):
                 file.write(string)
             self.openPositionRolling += 1
-            if self.openPositionRolling > 10:
+            if self.openPositionRolling > 3:
```
maybe have a constant somewhere, which is importable that everyone is able to define how much rolling position files they want to have?

Krysh90 commented on kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py:
```diff
@@ -8,13 +8,40 @@
+    def __init__(self, wsURLs, api_key, api_secret, logger, callback, on_message,
+                 symbol, minutesPerBar):
         self.messageCallback= on_message
+        self.minutesPerBar= minutesPerBar
```
minutes_per_bar

Krysh90 replied:
this property got changed in class below, but not in this?

Krysh90 commented on kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py:
```diff
@@ -8,13 +8,40 @@
         self.messageCallback= on_message
+        self.minutesPerBar= minutesPerBar
+        self.symbol= symbol
+        self.initialSubscribeDone= False
```
initial_subscribe_done

Krysh90 commented on kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py:
```diff
@@ -8,13 +8,40 @@
     def on_message(self, message):
         self.messageCallback(message)
 
+    def subscribeRealtimeData(self):
```
subscribe_realtime_data

Krysh90 commented on kuegi_bot/exchanges/bybit_linear/bybitlinear_websocket.py:
```diff
@@ -8,13 +8,40 @@
         self.messageCallback(message)
 
+    def subscribeRealtimeData(self):
+        subbarsIntervall = '1' if self.minutesPerBar <= 60 else '60'
```
subbars_interval

Krysh90 commented on kuegi_bot/utils/trading_classes.py:
```diff
@@ -182,14 +188,19 @@ def __init__(self, id: str, entry: float, stop: float, amount: float, tstamp):
         self.wanted_entry = entry
         self.initial_stop = stop
         self.amount = amount
+        self.maxFilledAmount= 0
```
max_filled_amount

Krysh90 commented on kuegi_bot/utils/trading_classes.py:
```diff
@@ -182,14 +188,19 @@ def __init__(self, id: str, entry: float, stop: float, amount: float, tstamp):
         self.initial_stop = stop
         self.amount = amount
+        self.maxFilledAmount= 0
+        self.currentOpenAmount= 0
```
current_open_amount

Krysh90 commented on kuegi_bot/utils/trading_classes.py:
```diff
@@ -182,14 +188,19 @@ def __init__(self, id: str, entry: float, stop: float, amount: float, tstamp):
         self.entry_tstamp = 0
         self.exit_tstamp = 0
         self.exit_equity = 0
+        self.customData= {}
```
custom_data

kuegi committed:
backtest: added error handling and fixed executions

kuegi committed:
backtest: fix order of execution

can't sort by distance to close, as close might jump into the middle of limits, need to execute limits in the correct order.

kuegi committed:
tradeengine: improved error logging

kuegi committed:
fix handling of connectedOrders

adapted kuegiStrat to use partial fills

kuegi committed:
fixed from live-test

kuegi committed:
dashboard: added currentOpenAmount

kuegi committed:
index: fix formatting

kuegi committed:
dashboard: improved price formatting

kuegi committed:
bybit_linear: don't use apikey for public. not necessary

kuegi committed:
refactor signal id

kuegi committed:
backtest: ensure correct handling of changed limits

can't execute changed orders the same as unchanged

kuegi committed:
exit: added TimedExitModule

kuegi committed:
MultiStrat: only giving each strat their open_positions

kuegi committed:
trading_bot: improve fixing of unaccounted position

kuegi committed:
[MultiStratBot] fix handling of openpositions

canceled positions were not canceled in the bot.

kuegi committed:
[bybit interface] fix stop_triggered

a stop_limit that is triggered in bybit is moved from a stop-order to a limit-order (so the new order has no stop anymore and we don't know that it was a triggered stopLimit before)-> need to set stop_triggered manually when a stop is gone from an order

kuegi committed:
backtestengine: fix taker/maker fee of instant stopLimit

kuegi committed:
Interfaces & TradingBot

fixed handling of executions, removed legacy handling if exchange can handle it already.
also fixed handling of discrepancies when race conditions on executions happen

kuegi committed:
correct way of handles_executions

Krysh90 committed:
[ kuegi ] added take profit with tp_fac (#15)

kuegi committed:
refactors

# Conflicts:
#	kuegi_bot/backtest_engine.py

kuegi committed:
even more refactors

kuegi committed:
moar refactor and fixes

kuegi committed:
fix refactor in interface

Krysh90 commented on kuegi_bot/backtest_engine.py:
```diff
@@ -93,6 +94,14 @@ def send_order(self, order: Order):
             return
+        [posId, order_type]= TradingBot.position_id_and_type_from_order_id(order.id)
+        if order_type == OrderType.ENTRY:
+            [posId, direction]= TradingBot.split_pos_Id(posId)
```
shouldn't it be called `split_pos_id`?

Krysh90 commented on kuegi_bot/backtest_engine.py:
```diff
@@ -170,45 +182,86 @@ def handle_order_execution(self, order: Order, intrabar: Bar):
             return 0
+        # sort buys after sells (higher number) when bar is falling
+        longFac= 1 if self.bars[0].close > self.bars[0].open else 2
+        shortFac= 1 if self.bars[0].close < self.bars[0].open else 2
```
long_fac and short_fac

Krysh90 commented on kuegi_bot/backtest_engine.py:
```diff
@@ -170,45 +182,86 @@ def handle_order_execution(self, order: Order, intrabar: Bar):
+                                order.amount < 0 and order.limit_price < intrabar_to_check.close)):
                             # close below/above limit: got definitly executed
                             should_execute= True
+                            force_taker= True # need to assume taker.
```
this was one of those changes, which makes backtest engine more reliable in sense of live executions?

kuegi replied:
yes

Krysh90 commented on kuegi_bot/bots/MultiStrategyBot.py:
```diff
@@ -67,7 +62,7 @@ def manage_open_order(self, order, position, bars, to_update, to_cancel, open_po
         pass
 
-    def open_orders(self, is_new_bar, directionFilter, bars, account, open_positions):
+    def open_orders(self, is_new_bar, directionFilter, bars, account, open_positions :dict):
```
minor code-styling: whitespace on wrong side of :

Krysh90 commented on kuegi_bot/bots/strategies/kuegi_strat.py:
```diff
@@ -34,29 +35,43 @@ def init(self, bars: List[Bar], account: Account, symbol: Symbol):
-            # added temporarily, cause sync with open orders is in the next loop and otherwise the orders vs
-            # position check fails
-            if order not in account.open_orders:  # outside world might have already added it
-                account.open_orders.append(order)
```
is this removal intended? maybe some unwanted change from PR #15 ?

kuegi replied:
intented, as this safety check is now done in trade_engine directly. -> no need to do it in every strat anymore

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -26,20 +26,20 @@ def __init__(self, logger, directionFilter: int = 0):
+        self.last_tick_time: datetime = None
         self.is_new_bar = True
-        self.open_positions = {}
+        self.open_positions : Dict[str,Position] = {}
```
minor code-styling: to be consistent with other properties in here. no leading whitespace before :

Krysh90 commented on kuegi_bot/bots/trading_bot.py:
```diff
@@ -167,50 +168,115 @@ def cancel_entry(self, positionId, account: Account):
+
+        # old way (fallback if executions wheren't there)
+        if not self.order_interface.handles_executions:
+            self.logger.warn("your exchange interface doesn't handle executions yet. please upgrade")
```
maybe this should be an error, to be more visible

kuegi committed:
tradingBot: fix error in pos handling

kuegi committed:
PR comments

kuegi committed:
reformat
